org.spoofax.interpreter.core/src/main/java/org/spoofax/interpreter/library/ssl/SSL_immutable_relation_transitive_closure.java

```diff
@@ -0,0 +1,55 @@
+package org.spoofax.interpreter.library.ssl;
+
+import io.usethesource.capsule.BinaryRelation;
+
+import org.spoofax.interpreter.core.IContext;
+import org.spoofax.interpreter.library.AbstractPrimitive;
+import org.spoofax.interpreter.stratego.Strategy;
+import org.spoofax.interpreter.terms.IStrategoTerm;
+import java.util.AbstractMap;
+import java.util.HashSet;
+import java.util.Map.Entry;
+
+public class SSL_immutable_relation_transitive_closure extends AbstractPrimitive {
+
+    protected SSL_immutable_relation_transitive_closure() {
+        super("SSL_immutable_relation_transitive_closure", 0, 0);
+    }
+
+    @Override
+    public boolean call(IContext env, Strategy[] sargs, IStrategoTerm[] targs) {
+        if(!(env.current() instanceof StrategoImmutableRelation)) {
+            return false;
+        }
+
+        final BinaryRelation.Immutable<IStrategoTerm, IStrategoTerm> map = ((StrategoImmutableRelation) env.current()).backingRelation;
+
+        env.setCurrent(new StrategoImmutableRelation(transitiveClosure(map)));
+        return true;
+    }
+
+    public static BinaryRelation.Immutable<IStrategoTerm, IStrategoTerm> transitiveClosure(
+        BinaryRelation.Immutable<IStrategoTerm, IStrategoTerm> map) {
+        HashSet<Entry<IStrategoTerm, IStrategoTerm>> frontier1 = new HashSet<>(map.entrySet());
+        HashSet<Entry<IStrategoTerm, IStrategoTerm>> frontier2 = new HashSet<>();
```

Comment:
I think the use of two sets here makes the implementation harder to understand. I normally use a single queue (linked list, I think), which has cheaper insert and remove than the sets.

Reply:
It took me a while to even figure out how to write this, so it's probably not the canonical implementation method. Do you have an nicer example implementation?

Reply:
I think I'd write something like this (note that I wrote this on Github):

    public static BinaryRelation.Immutable<IStrategoTerm, IStrategoTerm> transitiveClosure(
            BinaryRelation.Immutable<IStrategoTerm, IStrategoTerm> map) {
        final Deque<Entry<IStrategoTerm, IStrategoTerm>> worklist = new LinkedList<>(map.entrySet());
        final BinaryRelation.Transient<IStrategoTerm, IStrategoTerm> result = BinaryRelation.Transient.of();
        while(!worklist.isEmpty()) {
            final Entry<IStrategoTerm, IStrategoTerm> e = worklist.pop();
            for(IStrategoTerm pre : result.inverse().get(e.getKey())) {
                for(IStrategoTerm post : result.get(e.getValue())) {
                    if(!result.containsEntry(pre, post)) {
                        result.__insert(pre, post);
                        worklist.add(new AbstractMap.SimpleImmutableEntry<>(pre, post));
                    }
                }
            }
        }
        return result.freeze();
    }

Reply:
I copied that code, but it had a few bugs. I fixed them by making it looks more like my code. Then I created a benchmark with some really large graphs and noticed no performance difference. So I'll use the linked list, since it's slightly simpler. Thanks for the example code :)

Reply:
Cool! I always thought you'd have to do to full cross product (pre * post), but only halve seems to be enough. Also, that you only have to add entries based on the original relation instead of the new, closed result. I always find it hard to reason about what is enough to get the closure, and it seems less than I thought. I tested a few alternatives, but haven't been able to break your algorithm so far. 👍